### DISCUSSION TITLE
AttributeError: 'InstanceState' object has no attribute 'entity'

### DISCUSSION THREAD

**Discussion by abhi8585** [apache/airflow#24922] [Q&A]
Hi Community, I have upgraded my local airflow instance from version 1.10.5 to 2.3.2. I am using astronomer for managing my airflow instance. I have successfully completed the steps mentioned in the below link. There are no error logs in any of the docker containers of my airflow instance.
https://airflow.apache.org/docs/apache-airflow/stable/upgrading-from-1-10/index.html

At the time of loading DAGS in my airflow instance. I am getting an unusual error with sqlalchemy which states "AttributeError: 'InstanceState' object has no attribute 'entity'". For more information check the complete traceback of the error.

Broken DAG: [/usr/local/airflow/dags/dag_name.py] Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/util.py", line 503, in __init__
    self._aliased_insp = AliasedInsp(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/util.py", line 647, in __init__
    mapped_class_or_ac = inspected.entity
AttributeError: 'InstanceState' object has no attribute 'entity'

The DAG is responsible for the below tasks.
1. Create a presentation for a list of tables in my presentation cluster.
2. Load data for respective tables to the presentation cluster.

I have the below versions for packages related to sqlalchemy.
SQLAlchemy                               1.4.9
sqlalchemy-bigquery                      1.4.3
SQLAlchemy-JSONField                     1.0.0
sqlalchemy-redshift                      0.8.6
SQLAlchemy-Utils                         0.38.2

Any help would be appreciated.

Thx!

**Comment by potiuk** [comment]
I think this is something with your Dag, but hard to help you without seeing the DAG and complete stack trace (the fragment you've shown does not show where it came from).

**Comment by abhi8585** [comment]
Hi Community,

I have tried my best to explain my DAG functionality with some sample tables and variable names. The traceback that I can see
at the Airflow UI for the respective DAG leads toward the usage of aliased function in the table1.py file. I have looked into the
source code of the sqlalchemy orm module, the code in the traceback failed when it looks for the inspection of the Aliased class
at the time of inspection it is not able to find the attribute entity. When I comment the aliased method line in the code my DAG
loads successfully in the UI without any error.


    #dag_file.py
    
    from airflow import DAG
    from src.utils import get_variable_list
    from dags.common.variable_dag_generator import generate_dags
    from dags.common.create_dag import create_dag
    from folder.table1 import table1
    def create_dags(dag_id,
                   schedule,
                   dag_number,
                   pod_id,
                   connection_config,
                   additional_config,
                   default_dag_args):

    SCHEMA = connection_config['redshift_exchange_rate_schema']


    table_names = [{'name': 'table1',
                    'transformation_query': table1(schema=SCHEMA).query(session=session,schema=REDSHIFT_STAGING_SCHEMA)}]

    dag = DAG(dag_id,
              schedule_interval=schedule,
              default_args=default_dag_args,
              max_active_runs=1)

    return create_dag(dag, pod_id, connection_config, table_names)


    for k, v in generate_dags(get_variable_list("some_variable_name"), "some_dag_name",
                            create_dags).items():
        # Below line adds the dag returned for each pod to the global scope, so Airflow can recognise the Dags
        globals()[k] = v
    
      #table1.py
      
      from sqlalchemy import cast,func, Column, BIGINT, String, TIMESTAMP, Boolean, \
          ForeignKeyConstraint, PrimaryKeyConstraint, \
          literal_column, literal
      from sqlalchemy.orm import aliased
      from sqlalchemy_redshift.dialect import DOUBLE_PRECISION
      from sqlalchemy import and_
      from sqlalchemy import cast
      from tables.folder.table2 import table2
      from tables.folder.table3 import table3
  
  
  
      class table1(schema_name):
          col1 = Column(String(512))
          col2 = Column(BIGINT)
          col3 = Column(BIGINT)
          col4 = Column(String(512))
      
      
          def __source(self,schema,is_migration):
              if is_migration == False:
                  var1 = aliased(table2(schema=schema))
                  var2 = aliased(table2(schema=schema))
              else:
      
                  var1 = aliased(table3(schema=monetate_schema))
                  var2 = aliased(table3(schema=monetate_schema))
      
              query = "some select query"
              return query
      
  
      def query(self, session,schema,is_migration):
          return str(self.__source(session, schema_name).filter(table.updatedate >= func.dateadd(
                                                                              start_date)
                                        ).filter(table.updatedate < end_date
                                                     ).statement.compile(dialect=engine.dialect
                                                                         , compile_kwargs={'literal_binds': True}
                                                                         )).replace('\'', '\\\'')
  
  
        
        #table2.py
        
        from sqlalchemy import Column, BIGINT, String, TIMESTAMP
        from sqlalchemy_redshift.dialect import DOUBLE_PRECISION
        
        
        
        class table2(Base):
            __tablename__ = 'table2'
        
      col1 = Column(TIMESTAMP)
      col2 = Column(String(512), primary_key=True)
      __table_args__ = {}
  
      def __init__(self, schema=schema):
          self.__table__.schema = schema
  
  
      #table3.py
      
      from sqlalchemy import Column, BIGINT, String, TIMESTAMP
      from sqlalchemy_redshift.dialect import DOUBLE_PRECISION
      
      
      
      class table3(Base):
          __tablename__ = 'table3'
      
          col1 = Column(TIMESTAMP)
          col2 = Column(String(512), primary_key=True)
          __table_args__ = {}
      
          def __init__(self, schema=schema):
              self.__table__.schema = schema

**Reply by potiuk** [reply]
I think you should post your question with SQLalchemy  this error has nothing to do with Airflow. But yes we use newer version of sqlalchemy. You should find out with them how to update your code IMHO as you are using the ORM capabilities of it

**Reply by abhi8585** [reply]
Hi @potiuk , Which version of SQLalchemy is compatible with Airflow 2.3.2?

**Reply by potiuk** [reply]
Did you check what are the requirements of Airflow @abhi8585 ?  You know it's public info and you can find out more by reading https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html
